.github/workflows/ci.yml

```diff
@@ -160,13 +160,13 @@ jobs:
       - name: Create temp files and folders
         run: mkdir -p var/logs
 
+      - name: Build external services
+        run: yarn external-services-test build graph
+
       - name: Launch external services
-        if: ${{ success() || failure() }}
-        run: |
-          yarn external-services-test up --detach
+        run: yarn external-services-test up --detach
```

Comment:
I actually don't understand why we'd want to do this, I don't get how this could be faster, and will result in a _sequential_ build of the graph within external services and then the rest of the images

Reply:
As it turned out that we obviously had built the graph multiple times and this avoids building the graph multiple times by explicitly building it before. The image will then be re-used.

Reply:
Building and launching the external services in this run took about 7 minutes. In the previous PR it took 11 minutes.

Reply:
On the first run of a PR yesterday, before we started seeing weird CI behaviour, building external services took 7 minutes
https://github.com/hashintel/hash/actions/runs/4617340362/jobs/8163480386

Reply:
If you really want to test this, I'd suggest we wait for the PR with the fix (`locked`) to go in, see how long it takes a fresh commit on _another_ PR to build. And then open a PR with these changes again and see how much it improves on a fresh commit. 

So we run 0 risk of any caching, or inconsistencies between combinations of changes

Reply:
The latest run with the main branch merged has unchanged times: It's still around 7 minutes for both, building and running.

Reply:
The latest run with the main branch merged in is still faster than the previous run. It can also easily be seen, that launching did not build any Rust-related crate.

Reply:
[On the latest run](https://github.com/hashintel/hash/actions/runs/4630336434/jobs/8191779581) I'm seeing 
- Building and starting taking a combined time of almost 9 minutes on the _backend integration tests job_:
<img width="1310" alt="image" src="https://user-images.githubusercontent.com/25749103/230583714-b0477667-6d07-4f2d-8d0f-b739c10440d1.png">

- Building and starting taking a combined time of **almost 12 minutes** on the _Playwright job_:
<img width="1311" alt="image" src="https://user-images.githubusercontent.com/25749103/230583297-06ec2a37-cb45-459e-96bd-abc16c61df8a.png">


However, [on the fresh run I posted before](https://github.com/hashintel/hash/actions/runs/4617340362/jobs/8163480511) 
-**_7.5 minutes_** on the _backend integration tests job_
<img width="1320" alt="image" src="https://user-images.githubusercontent.com/25749103/230583875-81050b57-d746-4e5c-a56d-5e8908e77bdb.png">

- **_1 minute_** on the playwright job on the run I posted
<img width="1319" alt="image" src="https://user-images.githubusercontent.com/25749103/230583511-48379da8-8693-4041-bc3e-f9f06949fde3.png">

I do not understand how it is still faster. And I'd again like to suggest
> I don't think we have capacity to properly assess this, and I'd encourage us to leave it for now.

Reply:
The playwright tests are slower because, for whatever reason, there is a `--build` parameter passed to the `up` command (as previously).

Reply:
> 2 minutes on the playwright job on the run I posted

I don't see what this has to do with this PR. The linked Job is failing the launch because of the issue we solved yesterday

Reply:
> I don't think we have capacity to properly assess this, and I'd encourage us to leave it for now.

I highly disagree. This is a net win in CI times if docker for some reason tries to build the graph multiple times and the PR is up and ready to be merged. Docker tries to be smart and creates multiple services at once so it may build the graph twice. With this PR this is avoided by building the image in advance.

Reply:
> I don't see what this has to do with this PR. The linked Job is failing the launch because of the issue we solved yesterday

Sorry you are entirely correct and it was a silly comparison. I had tried to pick a job which had one commit (to remove the small chance of caching affecting things), but today I forgot that at the time I picked them we didn't have the fix merged into main.

I've compared to [a different job](https://github.com/hashintel/hash/actions/runs/4630488031/jobs/8192120094)  and the times are much more similar between this PR and the new one. Importantly I do not see any increases in performance with the changes, so I do not see a point in us adding them in, as I mentioned above.

> This is a net win in CI times if docker for some reason tries to build the graph multiple times and the PR is up and ready to be merged

I am not convinced by this

> With this PR this is avoided by building the image in advance.

As I mentioned, from my understanding this "will result in a sequential build of the graph within external services and then the rest of the images". So even if this speeds up the graph building a bit, we're still _then_ building the postgres images and such after this specific build step.

Reply:
> So even if this speeds up the graph building a bit, we're still then building the postgres images and such after this specific build step.

I forgot, that we're actually building our own postgres image/wrapper image. Thanks for being pedantic!

I guess we're going to change the behavior of this workflow in the near future anyway due to the utilization of Turborepo?

Reply:
> I guess we're going to change the behavior of this workflow in the near future anyway due to the utilization of Turborepo?

I'm unclear if that's the case, but I suspect we will be adjusting our infra/docker setup in the near future anyway